### PR TITLE
fix(threads): typed ThreadNotFound error + skip Sentry for stale-thread RPCs (OPENHUMAN-TAURI-4H, -60)

### PR DESCRIPTION
**Pull request by oxoxDev**
## Summary

- Adds a typed `ThreadsError::NotFound` path for stale conversation thread references.
- Normalizes missing-thread display text to `thread <id> not found`.
- Returns structured JSON-RPC error data with `kind: "ThreadNotFound"` and skips Sentry reporting for this expected user state.
- Teaches the app RPC client and thread store to clear stale thread state, refresh the thread list, and suppress user-facing error toast noise.
- Adds backend, JSON-RPC, Sentry transport, and frontend tests for stale-thread handling.

## Problem

- Stale tabs or deleted/expired thread IDs can cause `openhuman.threads_message_append` and `openhuman.threads_generate_title` to run against threads that no longer exist.
- These are expected user-state races, but they were reported through `report_error`, creating Sentry noise under `domain=rpc operation=invoke_method`.
- The backend emitted two string shapes (`does not exist` and `not found`), and the frontend had no structured discriminator to branch on safely.

## Solution

- Introduce `ThreadsError::NotFound { thread_id }` and canonical parsing for both current and legacy missing-thread text.
- Map stale thread failures from `threads_message_append` and `threads_generate_title` to the typed error.
- At the JSON-RPC boundary, classify `ThreadNotFound` only for `openhuman.threads_*` methods, return structured error data, and skip `report_error`.
- Add a Sentry `TestTransport` smoke test proving `ThreadNotFound` does not emit an event while unrelated RPC errors still do.
- Extend frontend RPC error taxonomy with `thread_not_found`; clear stale thread refs, reload threads, avoid retrying the failed RPC, and suppress the send-error toast for the benign stale-thread case.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) per [Testing Strategy](../gitbooks/developing/testing-strategy.md#failure-path-requirement)
- [x] **Diff coverage ≥ 80%** — not run locally (`pnpm test:coverage` / `pnpm test:rust` not run); CI `diff-cover` is the authoritative check.
- [x] Coverage matrix updated — N/A: no feature rows added, removed, or renamed.
- [x] All affected feature IDs from the matrix are listed in the PR description under `## Related` — N/A: Sentry-noise cleanup is not represented by a matrix feature ID.
- [x] No new external network dependencies introduced (mock backend used per [Testing Strategy](../gitbooks/developing/testing-strategy.md#mock-policy))
- [x] Manual smoke checklist updated if this touches release-cut surfaces — N/A: no release smoke checklist surface changed.
- [x] Linked issue closed via `Closes #NNN` in the `## Related` section — N/A: source task tracks Sentry issue IDs, not a GitHub issue number.

## Impact

- Desktop/app thread UX: stale thread references now disappear quietly instead of causing repeated failed sends/title-generation attempts or visible error toasts.
- Core observability: expected stale-thread RPC failures no longer report to Sentry; unrelated RPC failures still report normally.
- Compatibility: JSON-RPC error message remains human-readable while adding structured `data` for newer clients.
- No migration or external network dependency changes.

## Related

- Closes OPENHUMAN-TAURI-4H
- Closes OPENHUMAN-TAURI-60
- Follow-up PR(s)/TODOs: none

---

## AI Authored PR Metadata (required for Codex/Linear PRs)

> Keep this section for AI-authored PRs. For human-only PRs, mark each field `N/A`.

### Linear Issue
- Key: N/A
- URL: N/A

### Commit & Branch
- Branch: `fix/thread-not-found-noise`
- Commit SHA: `be9647f596d185a416edf69749232f7a5aa82f63`

### Validation Run
- [x] `pnpm --filter openhuman-app format:check` (also passed in pre-push hook)
- [x] `pnpm typecheck`
- [x] Focused tests:
  - `cargo test --lib thread_not_found_rpc_error -- --nocapture`
  - `cargo test --lib openhuman::threads`
  - `cargo test --test json_rpc_e2e`
  - `pnpm debug unit src/services/__tests__/coreRpcClient.test.ts src/store/__tests__/threadSlice.test.ts`
  - `pnpm --filter openhuman-app test:unit`
- [x] Rust fmt/check (if changed):
  - `cargo fmt --check --all`
  - `cargo check --manifest-path Cargo.toml`
  - `cargo test --lib openhuman::threads`
  - `cargo test --test json_rpc_e2e`
- [x] Tauri fmt/check (if changed):
  - `cargo fmt --manifest-path app/src-tauri/Cargo.toml --all --check`
  - `pnpm --filter openhuman-app rust:check`

### Validation Blocked
- `command:` `cargo clippy --workspace -- -D warnings`
- `error:` fails on pre-existing workspace warnings/errors outside this patch, including unused imports in `src/openhuman/agent/harness/builtin_definitions.rs`, `src/openhuman/channels/providers/telegram/channel.rs`, `src/openhuman/composio/providers/profile.rs`, `clippy::empty_line_after_doc_comments` in `src/openhuman/tools/impl/agent/onboarding_status.rs`, and many existing `derivable_impls` / `manual_pattern_char_comparison` / `too_many_arguments` findings.
- `impact:` workspace clippy remains blocked independently of this branch; changed behavior is covered by the focused Rust/JSON-RPC/frontend tests above.

### Behavior Changes
- Intended behavior change: missing thread RPCs are treated as expected stale-client state instead of reportable backend failures.
- User-visible effect: stale selected/active threads are cleared and the list is refreshed without a toast or retry loop.

### Parity Contract
- Legacy behavior preserved: unrelated RPC errors still call `report_error`; session-expired suppression remains unchanged.
- Guard/fallback/dispatch parity checks: JSON-RPC classification is scoped to `openhuman.threads_*` and rejects unrelated `not found` strings such as message-level misses.

### Duplicate / Superseded PR Handling
- Duplicate PR(s): N/A
- Canonical PR: this PR
- Resolution (closed/superseded/updated): N/A

## Summary by CodeRabbit

* **New Features**
  * Server now returns structured "thread not found" errors so the app can detect stale threads reliably.

* **Bug Fixes**
  * Stale threads are evicted automatically, selected/active threads cleared, thread list refreshed, and user-facing messages standardized to "thread ... not found" instead of generic send failures.

* **Tests**
  * Added unit and end-to-end coverage for thread-not-found handling and observability behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tinyhumansai/openhuman/pull/1570)

### CODE REVIEW

File: src/core/jsonrpc.rs [hunk 1]
```diff
@@ -55,19 +56,42 @@ pub async fn rpc_handler(State(state): State<AppState>, Json(req): Json<RpcReque
...
+            // set the `expected_user_state` flag on their structured envelope
+            // and skip Sentry here uniformly.
+            if expected_user_state {
+                tracing::info!(
+                    method = %method,
+                    "[rpc] expected-user-state error — skipping Sentry: {}",
+                    display_message
+                );
+            } else if !is_session_expired_error(&display_message) {
                 crate::core::observability::report_error(
```
Comment (graycyrus) [inline src/core/jsonrpc.rs]: **[major]** Potential merge-conflict regression: `report_error` vs `report_error_or_expected`

The current `main` branch uses `report_error_or_expected` at this call site (introduced to suppress Sentry noise for `LocalAiDisabled` and `ApiKeyMissing` conditions). This PR replaces it with bare `report_error`, which would re-introduce Sentry noise for those expected errors once merged.

This likely happened because the PR branched before `report_error_or_expected` landed on `main`. When rebasing, make sure this line becomes:

```rust
crate::core::observability::report_error_or_expected(
    display_message.as_str(),
    "rpc",
    "invoke_method",
    &[("method", method.as_str()), ("elapsed_ms", &ms.to_string())],
);
```

Both suppression mechanisms (structured `expected_user_state` flag for new domains + `expected_error_kind()` pattern matching for existing plain-string errors) need to coexist.

File: app/src/store/threadSlice.ts [hunk 2]
```diff
@@ -136,6 +139,20 @@ export const addMessageLocal = createAsyncThunk(
       }
       return { threadId: payload.threadId, message: persisted };
     } catch (error) {
+      if (isThreadNotFoundCoreRpcError(error, payload.threadId)) {
```
Comment (graycyrus) [inline app/src/store/threadSlice.ts:174]: **[minor]** Stale-thread recovery block is copy-pasted three times (here, `addInferenceResponse`, and `generateThreadTitleIfNeeded`).

Consider extracting to a shared helper:

```typescript
async function handleStaleThread(
  threadId: string,
  dispatch: AppDispatch,
  thunkName: string
): Promise<void> {
  dispatch(clearStaleThread(threadId));
  try {
    await dispatch(loadThreads()).unwrap();
  } catch (refreshError) {
    if (IS_DEV) {
      console.debug(`[threadSlice] ${thunkName} stale-thread refresh failed`, {
        threadId,
        error: refreshError,
      });
    }
  }
}
```

Functionally correct as-is — all three copies are tested — but DRY would reduce the chance of future divergence when the recovery logic evolves.

File: app/src/pages/Conversations.tsx [hunk 3]
```diff
@@ -581,6 +582,10 @@ const Conversations = ({ variant = 'page', composer = 'text' }: ConversationsPro
       await dispatch(addMessageLocal({ threadId: sendingThreadId, message: userMessage })).unwrap();
     } catch (error) {
       const msg = error instanceof Error ? error.message : String(error);
+      if (msg === THREAD_NOT_FOUND_MESSAGE) {
```
Comment (graycyrus) [inline app/src/pages/Conversations.tsx]: **[minor]** String equality check on `THREAD_NOT_FOUND_MESSAGE` works today but is fragile.

`dispatch(thunk).unwrap()` on a `rejectWithValue` throws the payload directly (a plain string), so `String(error)` matches. But if `addMessageLocal` ever wraps the rejection in an `Error` object (e.g. in a future middleware), this equality check silently stops matching.

A comment here would clarify the contract:

```typescript
// error is the rejectWithValue payload (plain string), not an Error object —
// THREAD_NOT_FOUND_MESSAGE is a string sentinel from threadSlice.
if (msg === THREAD_NOT_FOUND_MESSAGE) {
```

Alternatively, checking `isThreadNotFoundCoreRpcError` before `unwrap()` would be more robust, but that changes the control flow.

File: src/rpc/structured_error.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    #[serde(default)]
+    pub expected_user_state: bool,
+}
+
+impl StructuredRpcError {
+    /// Encode into a sentinel-prefixed string suitable for the controller
+    /// `Result<_, String>` error channel.
+    pub fn encode(&self) -> String {
+        // serde_json::to_string on a struct of String/Option<Value>/bool
+        // cannot fail in practice, so unwrap is acceptable.
```
Comment (graycyrus) [inline src/rpc/structured_error.rs:54]: **[nitpick]** `.unwrap_or_else(|_| String::from("{}"))` silently produces an invalid envelope if serialization fails. Since `serde_json::to_string` on `String + Option<Value> + bool` cannot fail in practice, consider `expect("StructuredRpcError serialization cannot fail")` — any future struct change that breaks serialization would then fail visibly in CI rather than producing a silent sentinel-prefixed `{}`.

File: tests/json_rpc_e2e.rs [hunk 5]
```diff
@@ -1019,6 +1019,69 @@ async fn json_rpc_thread_labels_create_and_update() {
...
+            "message": {
+                "id": "msg-1",
+                "content": "hello",
+                "type": "text",
+                "extraMetadata": {},
+                "sender": "user",
+                "createdAt": "2026-01-01T00:00:00Z"
+            }
+        }),
+    )
```
Comment (graycyrus) [inline tests/json_rpc_e2e.rs:1058]: **[nitpick]** The comment below says the transport no longer stamps `method` into the structured error data — but the test only asserts `kind` and `thread_id` are present without asserting `method` is *absent*. Adding:

```rust
assert!(
    append_err["data"]["method"].is_null(),
    "method must not appear in structured error data"
);
```

would make the contract explicit and catch a future accidental re-introduction.

**Review by graycyrus — COMMENTED** [review]
## Review — fix(threads): typed ThreadNotFound error + skip Sentry for stale-thread RPCs

### Walkthrough

Clean end-to-end implementation of typed stale-thread error handling. A generic `StructuredRpcError` envelope lets domain errors carry machine-readable `data` through the existing `Result<_, String>` channel without changing every controller signature. The transport layer decodes it domain-agnostically — no method-name branching. Frontend gets a `thread_not_found` error kind, clears stale state, and suppresses the toast. Test coverage is solid across all layers including a Sentry `TestTransport` smoke test.

### Changes

| File | Summary |
| --- | --- |
| `Cargo.toml` | Adds `"test"` feature to existing `sentry 0.47.0` |
| `src/rpc/structured_error.rs` | **New** — generic structured-error envelope with encode/decode |
| `src/openhuman/threads/error.rs` | **New** — `ThreadsError` enum with `NotFound` variant + parser |
| `src/openhuman/threads/ops.rs` | `message_append` and `thread_generate_title` return `ThreadsError` |
| `src/openhuman/memory/conversations/store.rs` | Normalises "does not exist" → "not found" |
| `src/core/jsonrpc.rs` | Transport decodes `StructuredRpcError`, skips Sentry for `expected_user_state` |
| `app/src/services/coreRpcClient.ts` | `thread_not_found` kind + `isThreadNotFoundCoreRpcError` helper |
| `app/src/store/threadSlice.ts` | `clearStaleThread` reducer + stale-thread recovery in 3 thunks |
| `app/src/pages/Conversations.tsx` | Suppresses send-error toast for stale threads |
| Tests (6 files) | Unit, JSON-RPC E2E, Sentry transport smoke, frontend slice + RPC client |

### Verified / looks good

- `StructuredRpcError::decode` is pure and panic-free — graceful `None` on corrupt envelopes
- `sentry "test"` feature is correctly scoped — not a new runtime dependency
- `clearStaleThread` reducer correctly handles `welcomeThreadId`
- `isThreadNotFoundCoreRpcError` with missing `errorThreadId` returns `true` — correct conservative behavior
- Generic test IDs (`"t-1"`, `"thread-123"`) — no real names/emails
- No `import.meta.env`, no `window.__TAURI__`, no dynamic imports in production code
- New Rust files placed in proper subdirectories, not at `src/openhuman/` root

See inline comments for actionable findings.
